src/packet.c

```diff
@@ -1233,35 +1220,29 @@ _libssh2_packet_requirev(LIBSSH2_SESSION *session,
         state->start = time(NULL);
     }
 
-    while (session->socket_state != LIBSSH2_SOCKET_DISCONNECTED) {
+    while (1) {
         int ret = _libssh2_transport_read(session);
-        if ((ret < 0) && (ret != LIBSSH2_ERROR_EAGAIN)) {
-            state->start = 0;
-            return ret;
-        }
-        if (ret <= 0) {
+        if (ret == LIBSSH2_ERROR_EAGAIN) {
             long left = LIBSSH2_READ_TIMEOUT -
                 (long)(time(NULL) - state->start);
 
             if (left <= 0) {
                 state->start = 0;
                 return LIBSSH2_ERROR_TIMEOUT;
             }
-            else if (ret == LIBSSH2_ERROR_EAGAIN) {
-                return ret;
-            }
+            return ret;
+        }
+        else if (ret < 0) {
+            state->start = 0;
+            return ret;
         }
 
-        if (strchr((char *) packet_types, ret)) {
+        if (ret && strchr((char *) packet_types, ret)) {
```

Comment:
Maybe add a comment here explaining that this will skip any packets with the invalid type 0